### DISCUSSION TITLE
日志文件为什么不支持添加路径

### DISCUSSION THREAD

**Discussion by zichengLeee** [chenshuo/muduo#585] [General]
为什么在LogFile的构造函数会
assert(basename.find('/') == string::npos);

**Comment by zichengLeee** [comment]
目前是不支持log文件 目录的自定义的吗？

**Comment by ender233** [comment]
官方给的例子默认这个地方就应该是一个文件名称而不是路径:
```
void setLogging(const char *argv0) {
    char name[256];
    strncpy(name, argv0, 256);
    g_asyncLog.reset(new muduo::AsyncLogging(::basename(name), kRollSize));
```
其中`::basename`就是返回路径的最后一个文件名:

> Return the file name within directory of FILENAME.  We don't
>    declare the function if the `basename' macro is available (defined
>    in <libgen.h>) which makes the XPG version of this function
>    available
  
基于此,当前的日志库是不支持目录自定义的,作者有一个[issue答复了这个](https://github.com/chenshuo/muduo/issues/191)

如果要插入路径那就修改代码, 应该在下面这个地方, `basename_`最终传给了`FileUtil::AppendFile`对象, 在这里打开了文件`FILE*`, 因此,可以在这里面打开文件的时候把路径放上去. 本身`fopen`就支持打开绝对或者相对路径的文件.
```
bool LogFile::rollFile()
{
  time_t now = 0;
  string filename = getLogFileName(basename_, &now);
  time_t start = now / kRollPerSeconds_ * kRollPerSeconds_;

  if (now > lastRoll_)
  {
    lastRoll_ = now;
    lastFlush_ = now;
    startOfPeriod_ = start;
    file_.reset(new FileUtil::AppendFile(filename));
    return true;
  }
  return false;
}
```
